test/disjointCoarseMallocPool.cpp

```diff
@@ -0,0 +1,507 @@
+                            &malloc_memory_provider);
+    ASSERT_NE(malloc_memory_provider, nullptr);
+
+    const size_t KB = 1024;
+    const size_t MB = 1024 * KB;
+
+    const size_t init_buffer_size = 20 * MB;
+
+    coarse_memory_provider_params_t coarse_memory_provider_params;
+    coarse_memory_provider_params.upstream_memory_provider =
+        malloc_memory_provider;
+    coarse_memory_provider_params.immediate_init = true;
+    coarse_memory_provider_params.init_buffer = NULL;
+    coarse_memory_provider_params.init_buffer_size = init_buffer_size;
+
+    umf_memory_provider_handle_t coarse_memory_provider;
+    umfMemoryProviderCreate(umfCoarseMemoryProviderOps(),
+                            &coarse_memory_provider_params,
+                            &coarse_memory_provider);
+    ASSERT_NE(coarse_memory_provider, nullptr);
+
+    umf_disjoint_pool_params_t disjoint_memory_pool_params = {};
+    disjoint_memory_pool_params.SlabMinSize = 4096;
+    disjoint_memory_pool_params.MaxPoolableSize = 4096;
+    disjoint_memory_pool_params.Capacity = 4;
+    disjoint_memory_pool_params.MinBucketSize = 64;
+    disjoint_memory_pool_params.PoolTrace = 1;
+
+    umf_memory_pool_handle_t pool;
+    umfPoolCreate(umfDisjointPoolOps(), coarse_memory_provider,
+                  &disjoint_memory_pool_params, UMF_POOL_CREATE_FLAG_NONE,
+                  &pool);
+    ASSERT_NE(pool, nullptr);
+
+    // test
+    double sizes[] = {2, 4, 0.5, 1, 8, 0.25};
+    size_t alignment[] = {0, 4, 0, 16, 32, 128};
+    for (int i = 0; i < 6; i++) {
+        size_t s = (size_t)(sizes[i] * MB);
+        void *t[8] = {0};
+        for (int j = 0; j < 8; j++) {
+            t[j] = umfPoolAlignedMalloc(pool, s, alignment[i]);
+            ASSERT_NE(t[j], nullptr);
+        }
+
+        for (int j = 0; j < 8; j++) {
+            umf_result_t res = umfPoolFree(pool, t[j]);
+            ASSERT_EQ(res, UMF_RESULT_SUCCESS);
+        }
+    }
+
+    umfPoolDestroy(pool);
+    umfMemoryProviderDestroy(coarse_memory_provider);
+    umfMemoryProviderDestroy(malloc_memory_provider);
+}
+
+struct alloc_ptr_size {
+    void *ptr;
+    size_t size;
+
+    bool operator<(const alloc_ptr_size &other) const {
+        if (ptr == other.ptr) {
+            return size < other.size;
+        }
+        return ptr < other.ptr;
+    }
+};
+
+TEST_F(test, disjointCoarseMMapPool_random) {
+
+    const size_t KB = 1024;
+    const size_t MB = 1024 * KB;
+    const size_t init_buffer_size = 200 * MB;
+
+    // Preallocate some memory
+    std::unique_ptr<char[]> buffer(new char[init_buffer_size]);
+    void *buf = buffer.get();
+    ASSERT_NE(buf, nullptr);
+    memset(buf, 0, init_buffer_size);
+
+    const unsigned char alloc_check_val = 11;
+
+    coarse_memory_provider_params_t coarse_memory_provider_params;
+    coarse_memory_provider_params.upstream_memory_provider = NULL;
+    coarse_memory_provider_params.immediate_init = false;
+    coarse_memory_provider_params.init_buffer = buf;
```

Comment:
Could you add a test case for both init_buffer and upstream_memory_provider set? What is the expected behavior in such a case?